feat: custom user translations

Shobhit-Nagpal opened the pull request:
## Changes made:
- created translations module on the backend to store user translations
- added a `Custom` option under the select for View languages in General Settings
- created a modal ui for the user to input their custom translations, english translations being the default fallback
- added additional logic on the general settings form's `onSubmit` as there was a behavior of the settings data being submitted along with the translations data when only clicked on saving the translations. this was due to nested forms, found a method to fix under [a RHF discussion](https://github.com/orgs/react-hook-form/discussions/7038#discussioncomment-11228568)
- transforming the translation object's keys as RHF was converting `common.expected_finish` to `common: { expected_finish }` when submitting the data
- refactored the `<TranslationProvider>` to include user translations
- added a `populateTranslation` function to create the translations files and folder on server startup

Attaching a video to show the working below:

https://github.com/user-attachments/assets/46e51294-9682-4923-ab1d-a03cca08cfde

Couple points to think about:
- Would like an opinion on the transformation of keys to prevent RHF to create objects from keys with dot notation
- Is there any better way of fetching data and providing it to the components than from the `<TranslationProvider>`? My initial thought was get the translations directly from the settings form but then thought it'd make more sense to access the functions through the translation context
- Have gone with a similar approach on the backend for translations as we did for custom CSS. The default translation is of English. When filling the form on the modal, no value can be empty either. Is this the flow we'd like to go for?
- Would also like some thoughts on the approach for `onSubmit` under General Settings now that there's nested forms into the picture. I did try using a React Portal but that did not seem to work. Curious if there's a better way to do this

If there's any change, do let me know

Shobhit-Nagpal commented on apps/server/src/api-data/translations/translations.service.ts:
```diff
@@ -0,0 +1,22 @@
+import { publicFiles } from '../../setup/index.js';
+import { defaultTranslation } from '../../user/translations/bundledTranslation.js';
+
+export async function readUserTranslation(): Promise<TranslationObject> {
```
i like the idea, making the change

Shobhit-Nagpal commented on apps/client/src/features/app-settings/panel/settings-panel/GeneralSettings.tsx:
```diff
@@ -39,8 +45,16 @@ export default function GeneralSettings() {
-  const onSubmit = async (formData: Settings) => {
+  const onSubmit = async (formData: Settings, event?: BaseSyntheticEvent) => {
     try {
+      // The target logic is to prevent the submission of the settings form from the translations form
```
i did try using `stopPropagation` before using react portals and it still didn't work. i'll check the custom css code once to see what we did there

Shobhit-Nagpal replied:
update on this: 

the issue is solved. i checked the css code again and turns out we're not using react hook form for the custom css. i tried the approach again with `stopPropagation` and portals as the react hook form discussion mentioned this approach but it seemed a bit complicated for something simple. ended up checking the code again and now the `<TranslationModal>` is being rendered outside the `<Panel.Section>` which is acting as the form so we don't have nested forms anymore

Shobhit-Nagpal commented:
hey, i've made the changes accordingly. do let me know if there's anything else to be done

cpvalente commented on apps/client/src/features/app-settings/panel/settings-panel/composite/CustomTranslationModal.tsx:
```diff
@@ -0,0 +1,353 @@
+  onClose: () => void;
+}
+
+export default function CustomTranslationModal({ isOpen, onClose }: CustomTranslationModalProps) {
```
question: can we make this interface dynamic to show all the properties in the translation object?

as of now, adding a single entry would mean changes in many parts of the code base, which sounds like something we may forget to do or that can get out of sync

cpvalente reviewed (commented):
The code seems to work well, thank you
I believe we should investigate whether the `CustomTranslationModal` could have a dynamic form.
Unless there is a good reason for us not to want to introduce that code (error prone or complex) I believe it will improve the maintainability of the feature

cpvalente commented:
@coderabbitai full review

Shobhit-Nagpal commented:
made the following changes:
- added a `useCustomTranslation` hook to use react query to fetch the translations and invalidating it on posting new translation
- translation modal now uses `useDisclosure` for its functionality
- the "Add translation" button is always in view now
- refactored mapping translation structure from categories to the initial fields themselves
- cleaned up comment on error handling

recording of working after these changes (p.s: you can skip the recording from 0:15 to 1:38, it's just filling in the translations):


https://github.com/user-attachments/assets/f873a143-d6c4-497d-b012-7b1a76e3eea9

cpvalente commented:
@coderabbitai full review

cpvalente reviewed (approved):
😃
